src/m365/spo/commands/folder/folder-sharinglink-add.ts

```diff
@@ -0,0 +1,193 @@
+class SpoFolderSharingLinkAddCommand extends SpoCommand {
+  private readonly allowedTypes: string[] = ['view', 'edit'];
+  private readonly allowedScopes: string[] = ['anonymous', 'organization', 'users'];
+
+  public get name(): string {
+    return commands.FOLDER_SHARINGLINK_ADD;
+  }
+
+  public get description(): string {
+    return 'Creates a new sharing link to a folder';
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+    this.#initValidators();
+    this.#initOptionSets();
+    this.#initTypes();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        folderId: typeof args.options.folderId !== 'undefined',
+        folderUrl: typeof args.options.folderUrl !== 'undefined',
+        type: typeof args.options.type !== 'undefined',
+        expirationDateTime: typeof args.options.expirationDateTime !== 'undefined',
+        scope: typeof args.options.scope !== 'undefined',
+        retainInheritiedPermissions: !!args.options.retainInheritiedPermissions,
+        recipients: typeof args.options.recipients !== 'undefined'
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '-u, --webUrl <webUrl>'
+      },
+      {
+        option: '--folderId [folderId]'
+      },
+      {
+        option: '--folderUrl [folderUrl]'
+      },
+      {
+        option: '--type <type>',
+        autocomplete: this.allowedTypes
+      },
+      {
+        option: '--expirationDateTime [expirationDateTime]'
+      },
+      {
+        option: '--scope [scope]',
+        autocomplete: this.allowedScopes
+      },
+      {
+        option: '--retainInheritedPermissions [retainInheritedPermissions]'
+      },
+      {
+        option: '--recipients [recipients]'
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        const isValidSharePointUrl: boolean | string = validation.isValidSharePointUrl(args.options.webUrl);
+        if (isValidSharePointUrl !== true) {
+          return isValidSharePointUrl;
+        }
+
+        if (args.options.folderId && !validation.isValidGuid(args.options.folderId)) {
+          return `${args.options.folderId} is not a valid GUID`;
+        }
+
+        if (args.options.type && !this.allowedTypes.some(type => type === args.options.type)) {
+          return `'${args.options.type}' is not a valid type. Allowed values are: ${this.allowedTypes.join(',')}`;
+        }
+
+        if (args.options.scope) {
+          if (!this.allowedScopes.includes(args.options.scope)) {
+            return `'${args.options.scope}' is not a valid scope. Allowed values are: ${this.allowedScopes.join(', ')}.`;
+          }
+          if (args.options.scope === 'users' && !args.options.recipients) {
+            return `The 'recipients' option is required when scope is set to 'users'.`;
+          }
```

Comment:
Let's add this as a runsWhen in an optionSet, this will enable it to be prompted for the end-user